akka-actor/src/main/scala/akka/pattern/RetrySettings.scala

```diff
@@ -0,0 +1,112 @@
+    shouldRetry: Throwable => Boolean = _ => true) {
+
+  override def toString: String = {
+    s"RetrySettings(attempts=$attempts, delayFunction=$delayFunction)"
+  }
+
+  def withDelay(delayFunction: Int => Option[FiniteDuration]): RetrySettings = {
+    copy(delayFunction = delayFunction)
+  }
+
+  def withDelayFunction(delayFunction: java.util.function.IntFunction[Optional[java.time.Duration]]): RetrySettings = {
+    val scalaFunc: Int => Option[FiniteDuration] = attempt => {
+      val javaDuration = delayFunction.apply(attempt)
+      if (javaDuration.isPresent)
+        Some(javaDuration.get.toScala)
+      else
+        None
+    }
+    withDelay(scalaFunc)
+  }
+
+  def withDecider(shouldRetry: Throwable => Boolean): RetrySettings = {
+    copy(shouldRetry = shouldRetry)
+  }
+
+  def withDeciderFunction(shouldRetry: java.util.function.Function[Throwable, Boolean]): RetrySettings = {
+    copy(shouldRetry = shouldRetry.apply)
+  }
+}
+
+object RetrySettings {
+
+  def apply(attempts: Int): RetrySettings = {
+    // Start with a reasonable minimum backoff (e.g., 100 milliseconds)
+    val minBackoff: FiniteDuration = 100.millis
+
+    // Max backoff increases with attempts, capped to a sensible upper limit (e.g., 1 minute)
+    val maxBackoff: FiniteDuration = {
+      val base = minBackoff * math.pow(2, attempts).toLong
+      base.min(1.minute)
+    }
+
+    // Random factor can scale slightly with attempts to add more jitter
+    val randomFactor: Double = 0.1 + (attempts * 0.05).min(1.0) // cap at 1.0
+
+    apply(attempts, minBackoff, maxBackoff, randomFactor)
+  }
+
+  def create(attempts: Int): RetrySettings = {
+    apply(attempts)
+  }
+
+  def apply(attempts: Int, fixedDelay: FiniteDuration): RetrySettings = {
+    RetrySettings(attempts, new FixedDelayFunction(fixedDelay))
+  }
+
+  def create(attempts: Int, fixedDelay: Duration): RetrySettings = {
+    apply(attempts, fixedDelay.toScala)
+  }
+
+  def apply(
+      attempts: Int,
+      minBackoff: FiniteDuration,
+      maxBackoff: FiniteDuration,
+      randomFactor: Double): RetrySettings =
+    new RetrySettings(attempts, new ExponentialBackoffFunction(minBackoff, maxBackoff, randomFactor))
+
+  def create(attempts: Int, minBackoff: Duration, maxBackoff: Duration, randomFactor: Double): RetrySettings = {
+    apply(attempts, minBackoff.toScala, maxBackoff.toScala, randomFactor)
+  }
+
+}
+
+final class ExponentialBackoffFunction(minBackoff: FiniteDuration, maxBackoff: FiniteDuration, randomFactor: Double)
+    extends Function1[Int, Option[FiniteDuration]] {
+
+  override def apply(attempt: Int): Option[FiniteDuration] = {
+    Some(calculateExponentialBackoffDelay(attempt, minBackoff, maxBackoff, randomFactor))
+  }
+
+  override def toString: String = {
+    s"Exponential(minBackoff=$minBackoff, maxBackoff=$maxBackoff, randomFactor=$randomFactor)"
+  }
+}
+
+final class FixedDelayFunction(delay: FiniteDuration) extends Function1[Int, Option[FiniteDuration]] {
```

Comment:
I don't think this is needed. For this simple case the ordinary `withDelay` function is enough.

Reply:
For the record it was probably added because I insisted on a single RetrySettings class over an ADT of different retry settings, and to regain some debuggability by having concrete function impls with `toString`s that would help get something useful out when logging/printing the retry settings (compared to an opaque lambda).

Reply:
it's private now